fdbserver/VersionedBTree.actor.cpp

```diff
@@ -3797,16 +3810,23 @@ class DWALPager final : public IPager2 {
 	ACTOR static Future<Void> clearRemapQueue_impl(DWALPager* self) {
 		// Wait for outstanding commit.
 		wait(self->commitFuture);
+		self->setOldestReadableVersion(self->getLastCommittedVersion());
+		wait(self->commit(self->getLastCommittedVersion() + 1));
+
+		wait(self->remapCleanupFuture);
+
+		// Set remap cleanup window to 0 to allow the remap queue to drain.
+		state int64_t remapCleanupWindowBytes = self->remapCleanupWindowBytes;
+		self->remapCleanupWindowBytes = 0;
 
-		// While the remap queue isn't empty, advance the commit version and oldest readable version
-		// by the remap cleanup window and commit
 		while (self->remapQueue.numEntries > 0) {
-			self->setOldestReadableVersion(self->getLastCommittedVersion());
-			wait(self->commit(self->getLastCommittedVersion() + self->remapCleanupWindow + 1));
+			self->remapCleanupFuture = remapCleanup(self);
```

Comment:
The original advance version / wait for commit loop is the correct thing to do here.  Nothing should call `remapCleanup()` other than `commit_impl()` which starts a new cleanup cycle at the end of each commit.  If that cycle were still running when this line is reached, the existing cycle would be cancelled and would leak whatever pages it was in the process of cleaning up because the queue pops would take effect but the `remapCleanup(item)` actors that were waiting would be cancelled.

The loop will let the current cleanup process finish using the old window size, then the next one will use the new window size.  After the loop, do 1 more commit after restoring the old window size.

Reply:
Is it true that after setting the window size to 0, if `remapCleanup()` didn't completely clear the queue, it would mean there's something committed? In that case wouldn't that commit bump the commit version, instead of we having to bump it here?

Reply:
updated, please check.